docs/endpoint-recovery.md

```diff
@@ -0,0 +1,434 @@
+   recovery was considered necessary
+
+For an application to invoke `.Recover` it must already consider the endpoint
+unresponsive, therefore it seems reasonable to assume it won't continue
+communicating with the endpoint unless the recovery succeeds. As such there's
+no action required when `.Connectivity` transitions to `Degraded`. However,
+if recovery succeeds, the transition of `.Connectivity` from `Degraded` to
+`Available` provides the signal to restart communicating with the endpoint.
+
+For applications in the second class it is likely the case that they
+haven't themselves invoked `.Recover` because they are yet to observe a
+communication failure with the endpoint. As such there's also no requirement
+for action when `.Connectivity` transitions to `Degraded` on behalf of another
+application. However, it may be the case that communication failures are
+subsequently observed. It's not necessary to invoke `.Recover` if
+`.Connectivity` is already in `Degraded`, though there should also be no
+significant consequences if it occurs. If no communication failures are observed
+while `.Connectivity` is in `Degraded` then there's also no action required if
+it transitions to `Available`.
+
+### `mctpd` behaviours over an endpoint recovery sequence
+
+For hotplug devices recovering from a reset `Get Endpoint ID` should yield an
+EID of 0.
+
+Note that while we're under the `Treclaim` period the device's allocated EID has
+not been quarantined to the reuse FIFO, and is also not a member of the dynamic
+allocation pool. For the purposes of allocation, the device is still assigned
+the EID it had prior to taking a reset. If a device at the physical address of
+the device in `Degraded` responds to `Get Endpoint ID` with an EID of 0 inside
+the `Treclaim` period and has a UUID that matches the device in `Degraded`, then
+it should be reassigned its allocated EID. At the end of this sequence the
+reassignment should not be observable to MCTP-application-layer beyond the
+endpoint's `.Connectivity` member transitioning from `Degraded` to `Available`.
+
+## MCTP device lifecycle and consuming events from `mctpd` D-Bus objects
+
+The general strategy for tracking endpoint lifecycles is [as follows]
+[dbus-spec-standard-interfaces-objectmanager]:
+
+[dbus-spec-standard-interfaces-objectmanager]: https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager
+
+1. Add a match of the following form for event-driven updates to local
+   data-structures
+
+   ```
+   type='signal',sender='xyz.openbmc_project.MCTP',path_namespace='/xyz/openbmc_project/mctp'
+   ```
+
+2. Issue a call to `GetManagedObjects` `org.freedesktop.DBus.ObjectManager` on
+   `/xyz/openbmc_project/mctp` for initial population of local data-structures
+
+3. Issue
+
+### An example order of MCTP device events, consumed by `nvmesensor`
+
+1. A series of events yields a new FRU EEPROM device on an I2C bus already
```

Comment:
The new FRU EEPROM of the device on I2C bus some time does not mean the device is ready for the mctp communication.
For the card which have the embedded OS, the FRU EEPROM will available before the OS is ready to response for MCTP messages.

Reply:
It's up to external processes to determine when the endpoint is ready (and call SetupEndpoint). In some cases, the presense of the FRU EEPROM is enough. In others it may not be.

I have some plans to add SMBus ARP support to do the discovery, but that's a bit down the pipeline. That would be a different instance of those "external processes".

Reply:
`It's up to external processes to determine when the endpoint is ready (and call SetupEndpoint). In some cases, the presense of the FRU EEPROM is enough. In others it may not be.` That means we will have some device which is not ready for MCTP message before step 3. The failure in steps 3 will cause the added EID will removed from the D-Bus and the EID will never be added again.

Because in this case step 1 and 2 are from out side service, maybe we can add one note to request the service make sure that the device is ready for MCTP interface before call `setupEndpoint`.

Reply:
> Because in this case step 1 and 2 are from out side service, maybe we can add one note to request the service make sure that the device is ready for MCTP interface before call setupEndpoint.

I don't think there should be any mystery to this. Applications that have received some assertion that an MCTP-capable device exists can call `SetupEndpoint` multiple times until it succeeds. It's an error to assume that `SetupEndpoint` will always succeed, just like it's an error to assume that if `SetupEndpoint` succeeds that the endpoint object will always exist. I don't see a need for another mechanism to determine whether the endpoint is ready to communicate via MCTP (unless the device has implementation bugs?).